filebeat/reader/line/decoder.go

```diff
@@ -0,0 +1,166 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package line
+
+import (
+	"fmt"
+	"io"
+
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/transform"
+
+	"github.com/elastic/beats/libbeat/common/streambuf"
+)
+
+type decoderReader struct {
+	in         io.Reader
+	decoder    transform.Transformer
+	buf        *streambuf.Buffer
+	encodedBuf *streambuf.Buffer
+	bufferSize int
+	symlen     []uint8
+}
+
+func newDecoderReader(in io.Reader, codec encoding.Encoding, bufferSize int) *decoderReader {
+	return &decoderReader{
+		in:         in,
+		decoder:    codec.NewDecoder(),
+		buf:        streambuf.New(nil),
+		encodedBuf: streambuf.New(nil),
+		bufferSize: bufferSize,
+	}
+}
+
+func (r *decoderReader) read(buf []byte) (int, error) {
+	b := make([]byte, r.bufferSize)
+
+	if r.buf.Len() != 0 {
+		return r.copyToOut(buf)
+	}
+
+	for {
+		start := r.encodedBuf.Len()
+		n, err := r.in.Read(b[start:])
+		if err != nil {
+			return len(buf[:n]), err
+		}
```

Comment:
this still looks like a potential bug. 

Let's assume before entering the loop we have this initial state:

```
encodedBuf.Len = 0
file-read-offset: 0
r.buffer_size: 1000 // len(b) == 1000
b[0:1000] = "\0..."
buf[0:1000] = "\0..."
```

The file state will contain 26 charatecters:
```
"abcdefghijklmnopqrstuvwxyz"
```

Given this state we already 

Now the read fails, but some bytes have been copied into the "internal" buffer b. Return values:
```
n = 10
err = "ups"
```
The new state is:

```
encodedBuf.Len = 0
file-read-offset: 10
b[0:10] = "abcdefghij"
b[10:1000] = "\0..."
buf[0:1000] = "\0..."
```

Next calling 'return' with this state we tell the caller `n` bytes have been written to buf -> return values are:

```
n = 10
err = "ups"
buf[0:10] = "\0..." // <- bug 1, we return contents, without having copied anythin
```

That is no content has been copied at all, yet we tell the next reader in the chain to process 10 characters.

Worse, if the next reader tries to read more, we have the initial state for the current function call at:

```
encodedBuf.Len = 0
file-read-offset: 10
r.buffer_size: 1000 // len(b) == 1000
b[0:1000] = "\0..."
buf[0:1000] = "\0..."
```

This time the read succeeds. As the filter read pointer is already at offset 16, we will read 16 bytes only: ->

```
n = 16
err = 0
b[0:16] = "klmnopqrstuvwxyz"
```

Assuming we can convert by just copying (all characters in lower ASCII range) the function return will be:

```
n = 16
err = nil
buf[0:16] = "klmnopqrstuvwxyz"
```

Due to the error handling we made the callee process the result byte sequence:
`"\0\0\0\0\0\0\0\0\0\0klmnopqrstuvwxyz"`, but the expected byte sequence should have been `"abcdefghijklmnopqrstuvwxyz"`.


Another potential bug arises we run the loop twice, with the first read succeeding and the second failing. Assuming `n=5` in the first run and `n=2` in the second run (plus error), then we will return `n=2, err="ups", buf[0:2]="ab"`. The other 5 bytes that have been writte from the file are ignored.